### DISCUSSION TITLE
ERR_SERVER_ALREADY_LISTEN

### DISCUSSION THREAD

**Discussion by Tharun-Findhope** [vercel/community#2939] [Help] [answered]
### Summary

I have deployed a serverless function which is working nice since last month. Yesterday, I just added few metadata to the repo and, its showing this serverless function has crashed. I have already checked, there is no multiple instance of ``app.listen()`` or ``server.listen()`` as I had only added few metadata and didn't changed any route or code segment that may impact.

Yet, last deployment is working fine with its unique url, when I tried to make it production, still the issue persists. I even tried Redeploy-ing in case build cache was causing that. But, still its showing the same thing.

Any help would be appreciated.

### Steps to Reproduce

Error Logs
``
undefined ERROR
Unexpected error booting Vercel Runtime: Error: Listen method has been called more than once without closing.
    at new NodeError (node:internal/errors:399:5)
    at Server.listen (node:net:1827:11)
    at Bridge.listen (/var/task/___vc/chunk-MOXFNKOD.js:700:19)
    at attachServer (/var/task/___vc/__launcher.js:22:18)
    at /var/task/___vc/__launcher.js:194:21 {
  code: 'ERR_SERVER_ALREADY_LISTEN'
}
RequestId: 22cb2ddd-9408-4ea5-ab97-66ca4f0f8ec5 Error: Runtime exited with error: exit status 1
Runtime.ExitError
``

![image](https://github.com/vercel/community/assets/125960332/838a0641-aaac-4a31-9cba-e4fe59695c44)

**Comment by gunnarvelle** [comment]
We got the same issue. Last functioning deploy for my project was yesterday june 20, 09.08 CET. All deployments after 10.21 CET have failed with this message.

**Reply by gunnarvelle** [reply, marked as answer]
We managed to code around the problem. I still belive the reason it started to fail is because of a change in Vercel, but now we have a working fix.

Our apps use express as a server, but to make them work in vercel we explicitly do not call listen.
Old code:
```
export default express()
  .use((req, res) => app.handle(req, res))
  .listen(config.port, function (err) {
    if (err) {
      console.error(err);
      return;
    }
    console.log(`> Started on port ${config.port}`);
  });
```

New functioning code:
```
if (!config.isVercel) {
  app.listen(config.port, (err) => {
    if (err) {
      console.error(err);
      return;
    }
    console.log(`> Started on port ${config.port}`);
  });
}

export default app;
```
In addition we updated our vercel.json with `"IS_VERCEL": "true"`.

**Comment by amyegan** [comment]
@Tharun-Findhope @gunnarvelle Sorry you're running into this error. I have a few questions to help with bug investigation:

- What framework or language is your project using?
- Aside from the main framework, what packages are you using in the serverless function?
- Do you have a public repo or minimal reproduction that you can share?
- Can you share an example deployment unique url, custom domain, and/or project id to help us locate more error details?

**Reply by gunnarvelle** [reply]
We have several applications with this error, but I'll focus on [this](https://github.com/NDLANO/editorial-frontend). This is a React project written in TypeScript and js, running on Express and built with Webpack.

https://editorial-frontend-ocz99dld9-ndlaknowit.vercel.app/ is the last functioning build from this application, deployed 09:08 CET june 20th. 

https://editorial-frontend-ad8ny83pu-ndlaknowit.vercel.app/ have the error, deployed 11.55 CET the same day.
This is from [merging a pr](https://github.com/NDLANO/editorial-frontend/commit/a526e534532f1db4c876e4ff9a44ab49e2af462e) into master.

https://editorial-frontend-ouvwu2ryb-ndlaknowit.vercel.app/ is a functional application, deployed june 19th 14:31 CET and is from merging [another pr](https://github.com/NDLANO/editorial-frontend/commit/4bf94ad45029f9cf8fafa347295ec9f80347d0f5) into master.

The differences in code between these last two deployments is minimal and should not make the function to crash.

Thanks for looking into this!

**Reply by Tharun-Findhope** [reply]
It is basically a Node.js project written in TypeScript using npm. Some major libraries include express, node-fetch, winston for logging.  It is not a public repo, so I am not able to show you. But, I assure that changes I made was only in metadata and should not be concerning. 

Btw you can look into https://khushi-backend-5xb3jhtud-tharun-findhope.vercel.app/, this has a few metadata changes and first one to crash.

Thanks for your response and giving time.

**Reply by amyegan** [reply]
Do you see the same error if you [redeploy](https://vercel.com/docs/concepts/deployments/managing-deployments#redeploy-a-project) the earlier deployment that is still working?

This kind of problem is sometimes related to minor package updates. Does it work if you use Vercel CLI to [deploy from local build](https://vercel.com/docs/cli/deploying-from-cli#deploying-from-local-build-prebuilt)?

**Reply by Tharun-Findhope** [reply]
Yes, redeploying the earlier deployment that is still working by removing cache, still cause same error.

**Reply by gunnarvelle** [reply]
Same here. I checked out the commit with the functioning build and ran a manual deployment with cli. Got [this deployment](https://editorial-frontend-5vc7y6jdk-ndlaknowit.vercel.app). Still crashing.

**Reply by gunnarvelle** [reply]
Any chance the [upgrading of the edge runtime](https://github.com/vercel/vercel/commit/8b3a4146af68d2b7288c80a5b919d832dba929b5) is relevant here?

**Reply by amyegan** [reply]
Seems possible. You can try running a deployment with an older CLI version by setting an environment variable like `VERCEL_CLI_VERSION=vercel@30.2.2`. If this workaround solves it, we'll have a better idea of what may be causing the problem

**Reply by Jonas-C** [reply]
Hi, stepping in for @gunnarvelle! We've tried a couple of versions ranging from 27.3.0 up to 30.0.0 to no avail. I'll add some additional context as well:

We're running React 17 with express with a custom server-side rendering setup, with slight tweaks to account for differences in every application. The entry point exposes an express server that listens to port 3000. The entire build ends up in a `build` folder. Our vercel config uses the `builds` field to host our client bundle (and public assets) through `@vercel/static`, whereas our server (build/server.js) uses `@vercel/node`. As @gunnarvelle previously stated, this error occurred across all our applications without any major changes to dependencies or setup. Our storybook continues to work just fine with our current setup, but all of our server-side rendered applications are broken.

Including all of the packages we're using might be slightly overkill, but I'll compile a list and send it to you if need be! Is there any other information I can provide?
